clippy_lints/src/methods/filter_next.rs

```diff
@@ -1,13 +1,29 @@
-use clippy_utils::diagnostics::{span_lint, span_lint_and_sugg};
+use clippy_utils::diagnostics::{span_lint, span_lint_and_then};
 use clippy_utils::source::snippet;
 use clippy_utils::ty::implements_trait;
+use rustc_ast::{BindingAnnotation, Mutability};
 use rustc_errors::Applicability;
 use rustc_hir as hir;
 use rustc_lint::LateContext;
 use rustc_span::sym;
 
 use super::FILTER_NEXT;
 
+fn path_to_local(expr: &hir::Expr<'_>) -> Option<hir::HirId> {
```

Comment:
We have a `path_to_local` helper in `clippy_utils`. Can we use that here?

Reply:
If I remember correctly, `path_to_local` doesn't "work" if the local is reached through field accesses, so for example it would return `false` for `a.b[0]` if given the id of the local `a`, but we need it to return `true`, because we use this logic to figure out if a binding needs to be made mutable, and `a.b[0].find_mut(..)` requires `a` be mutable. Maybe this could be better named or have a comment explaining it.

Reply:
`clippy_utils::path_to_local` is too primitive in my opinion. It only handles the case where the expression is a Path, and its Res is a `Res::Local`. Maybe we should modify it and make it handle this new cases.

[(`path_to_local`'s source)](https://doc.rust-lang.org/beta/nightly-rustc/src/clippy_utils/lib.rs.html#450-457)